kernel/arch/dreamcast/kernel/irq.c

```diff
@@ -38,22 +38,102 @@ struct trapa_cb {
     void         *data;
 };
 
+/* Linked list of IRQ states, one is pushed onto the stack
+   every time the top-level ISR is entered. */
+struct irq_state {            // SIZE
+    bool           handled;   // 1 byte  /* mov.b only has 15-byte displacement */
+    uint8_t        code;      // 1 byte
+    uint16_t       evt;       // 2 bytes
+    volatile struct
+        irq_state *previous;  // 4 bytes
+};                            // 8 BYTES TOTAL
+
 /* Individual exception handlers */
 static struct irq_cb   irq_handlers[0x100];
-/* TRAPA exception handlers */
+/* TRAPA exception handlers. */
 static struct trapa_cb trapa_handlers[0x100];
-
-/* Global exception handler -- hook this if you want to get each and every
-   exception; you might get more than you bargained for, but it can be useful. */
+/* Global exception handler */
 static struct irq_cb   global_irq_handler;
 
 /* Default IRQ context location */
-static irq_context_t    irq_context_default;
+static irq_context_t   irq_context_default;
+/* Current IRQ state linked list pointer */
+static volatile struct irq_state *        /* Points to most recent IRQ state */
+       volatile        irq_state_current; /* Volatile pointer to volatile struct */
+
+/* Called when entiringan IRQ to push its state onto the stack as current. */
+inline static void irq_state_push(struct irq_state *current) {
+    current->previous = irq_state_current;
+    irq_state_current = current;
+}
+
+/* Called when exiting an IRQ to pop its state from the stack. */
+inline static void irq_state_pop(void) {
+    assert(irq_state_current);
+    irq_state_current = irq_state_current->previous;
+}
+
+/* Called to grab the IRQ state at a particular stack level. */
+inline static volatile struct irq_state *irq_state_n(size_t level) {
+    volatile struct irq_state *state = irq_state_current;
+    
+    for(size_t depth = 0; depth < level; ++depth) {
+        if(!state) break;
+        state = state->previous;
+    }
+
+    return state;
+}
+
+/* How deeply nested in IRQ calls are we? */
+size_t irq_int_depth(void) {
+    size_t depth = 0;
+    const irq_mask_t imask = irq_disable();
+
+    const volatile struct irq_state *state = irq_state_current;
+    
+    while(state) {
+        ++depth;
+        state = state->previous;
+    }
+
+    irq_restore(imask);
+
+    return depth;
+}
 
 /* Are we inside an interrupt? */
-static int inside_int;
-int irq_inside_int(void) {
-    return inside_int;
+bool irq_inside_int(void) {
+    return !!irq_state_current;
+}
+
+/* What's the active IRQ? */
+irq_t irq_active_int(size_t level) {
+    const irq_mask_t imask = irq_disable();
+
+    const volatile struct irq_state *state = irq_state_n(level);
+    const irq_t irq = state? state->evt : 0;
+
+    irq_restore(imask);
+    return irq;
+}
+
+/* Have we handled the active interrupt? */
+bool irq_handled_int(size_t level) {
```

Comment:
When or how would this be used? It seems as though the dispatcher is manually checking the handled state at each layer, so it wouldn't be to know if you should chain then. Would it be something like a nested IRQ's handler determining how to behave based on if the handler it is interrupting has finished yet or not?

At first glance I thought 'sure, API, expose that' but I'm just not clear on how you'd use it. It doesn't seem (which was another first thought) that it is meant to allow multiple handlers to run on the same IRQ and behave differently if one before handled it, since the dispatcher gates propagation based on handling (or maybe I'm misunderstanding that?). Then I thought 'maybe for a handler to have multiple paths based on if the IRQ it's handling has been handled?' but it would be able to track that locally.

Reply:
So this was really one of those things where I do believe 99% of the time this function will not be useful; however, I do think it's worth having for the sake of API completeness, and because I do think there could eventually be a few very niche scenarios where the user may need to know whether the exception was accepted or rejected.

For instance, perhaps the "handler" thing they have is also propagating something through multiple functions or multiple regions of code which could accept or deny, and the top-level dispatcher on the user-side might need to know which state the IRQ is in: handled or not.

or perhaps whatever is installed to handle the exception is wanting to check whether the exception type is one which is handled by default or not, and do something different accordingly.